File: app/static/js/sales.js
// static/js/sales.js

let cart = [];
let totalPrice = 0;

// Initialize SocketIO client
const socket = io();

// Handle stock updates
socket.on('stock_updated', (data) => {
    // Update product stock in the UI
    updateProductStockUI(data.id, data.stock);
});

socket.on('low_stock_alert', (data) => {
    alert(`Low stock alert for ${data.product_name}: Only ${data.stock} left!`);
});

// Function to filter products by category
function filterCategory(categoryId) {
    fetch(`/stock/categories/${categoryId}/products`)
        .then(response => response.json())
        .then(data => {
            const productList = document.getElementById('product_list');
            productList.innerHTML = '';

            data.products.forEach(product => {
                const listItem = document.createElement('li');
                listItem.id = `product-${product.id}`;
                listItem.innerHTML = `
                    <button onclick="addToCart(${product.id}, '${product.name}', ${product.price})" ${product.stock <= 0 ? 'disabled' : ''}>
                        ${product.name} - Ksh ${product.price} (${product.stock} left)
                    </button>
                `;
                productList.appendChild(listItem);
            });
        });
}

// Function to add product to cart
function addToCart(productId, productName, price) {
    fetch('/sales/add_to_cart', {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ product_id: productId, quantity: 1 })
    })
    .then(response => response.json())
    .then(data => {
        if (data.success) {
            const existingItem = cart.find(item => item.product_id === productId);
            if (existingItem) {
                existingItem.quantity += 1;
                existingItem.total_price += price;
            } else {
                cart.push({
                    product_id: productId,
                    product_name: productName,
                    quantity: 1,
                    price: price,
                    total_price: price
                });
            }
            updateCartUI();
        } else {
            alert(data.message);
        }
    })
    .catch(error => console.error('Error:', error));
}

// Function to update the cart UI
function updateCartUI() {
    const cartList = document.getElementById('cart_list');
    cartList.innerHTML = '';
    totalPrice = 0;

    cart.forEach(item => {
        const listItem = document.createElement('li');
        listItem.textContent = `${item.product_name} - Quantity: ${item.quantity} - Price: Ksh ${item.total_price}`;
        cartList.appendChild(listItem);
        totalPrice += item.total_price;
    });

    document.getElementById('total_price').textContent = totalPrice;
}

// Function to toggle customer name input based on payment method
function toggleCustomerName() {
    const paymentMethod = document.getElementById('payment_method').value;
    const customerNameDiv = document.getElementById('customer_name_div');
    if (paymentMethod === 'credit') {
        customerNameDiv.style.display = 'block';
    } else {
        customerNameDiv.style.display = 'none';
    }
}

// Function to complete the sale
function checkout() {
    if (cart.length === 0) {
        alert('Your cart is empty!');
        return;
    }

    const paymentMethod = document.getElementById('payment_method').value;
    const customerName = document.getElementById('customer_name').value || null;

    fetch('/sales/checkout', {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
            cart: cart,
            payment_method: paymentMethod,
            customer_name: customerName
        })
    })
    .then(response => response.json())
    .then(data => {
        if (data.success) {
            alert(data.message);
            cart = [];
            updateCartUI();
            // Optionally, reload products to reflect updated stock
            window.location.reload();
        } else {
            alert(data.message);
        }
    })
    .catch(error => console.error('Error:', error));
}

// Function to update product stock in the UI
function updateProductStockUI(productId, newStock) {
    const productItem = document.getElementById(`product-${productId}`);
    if (productItem) {
        const button = productItem.querySelector('button');
        button.innerHTML = button.innerHTML.replace(/\(\d+ left\)/, `(${newStock} left)`);
        button.disabled = newStock <= 0;
    }
}

File: app/sales.py
from flask import Blueprint, render_template, request, jsonify, flash, redirect, url_for
from flask_login import login_required, current_user
from app.models import db, Product, Sale, CartItem, Category
from app import socketio, limiter
from flask_socketio import emit
from datetime import datetime, timedelta
from sqlalchemy.exc import IntegrityError

sales_bp = Blueprint('sales', __name__)

# Helper function to check low stock
def check_low_stock(product):
    return product.stock < 5

# Route for cashier to view sales screen
@sales_bp.route('/sales')
@limiter.limit("200 per day")
@login_required
def sales_screen():
    categories = Category.query.all()
    return render_template('sales.html', categories=categories)

# API to fetch products by category
@sales_bp.route('/api/products/<int:category_id>', methods=['GET'])
@limiter.limit("200 per day")
@login_required
def get_products_by_category(category_id):
    products = Product.query.filter_by(category_id=category_id).all()
    product_list = [{'id': product.id, 'name': product.name, 'price': product.price, 'stock': product.stock} for product in products]
    return jsonify({'products': product_list})

# API to add item to cart (with quantity increment on repeated clicks)
@sales_bp.route('/add_to_cart', methods=['POST'])
@limiter.limit("500 per day")
@login_required
def add_to_cart():
    data = request.json
    product_id = data.get('product_id')
    quantity = data.get('quantity', 1)

    product = Product.query.get(product_id)
    if not product:
        return jsonify({'success': False, 'message': 'Product not found'}), 404

    if product.stock < quantity:
        return jsonify({'success': False, 'message': 'Insufficient stock'}), 400

    return jsonify({
        'success': True,
        'product_id': product.id,
        'product_name': product.name,
        'quantity': quantity,
        'price': product.price,
        'total_price': product.price * quantity
    })

# API to handle checkout
@sales_bp.route('/checkout', methods=['POST'])
@limiter.limit("100 per hour")
@login_required
def checkout():
    data = request.json
    cart = data.get('cart', [])
    payment_method = data.get('payment_method', 'cash')
    customer_name = data.get('customer_name')

    if not cart:
        return jsonify({'success': False, 'message': 'Cart is empty'}), 400

    total_amount = 0
    for item in cart:
        product = Product.query.get(item['id'])  # Changed from 'product_id' to 'id'
        if product is None:
            return jsonify({'success': False, 'message': 'Product not found'}), 404
        total_amount += product.price * item['quantity']

    sale = Sale(date=datetime.utcnow(), total=total_amount, payment_method=payment_method,
                customer_name=customer_name if payment_method == 'credit' else None)
    db.session.add(sale)

    try:
        for item in cart:
            product = Product.query.get(item['id'])
            if product.stock < item['quantity']:
                db.session.rollback()
                return jsonify({'success': False, 'message': f'Insufficient stock for {product.name}'}), 400

            product.stock -= item['quantity']
            # No commit here yet, we will do it after the loop

            cart_item = CartItem(product_id=product.id, quantity=item['quantity'], sale_id=sale.id)
            db.session.add(cart_item)

        db.session.commit()  # Commit all changes at once

        # Emit real-time updates after successful commit
        for item in cart:
            product = Product.query.get(item['id'])
            socketio.emit('stock_updated', {'id': product.id, 'name': product.name, 'stock': product.stock}, broadcast=True)
            if check_low_stock(product):
                socketio.emit('low_stock_alert', {'product_name': product.name, 'stock': product.stock}, broadcast=True)

        return jsonify({'success': True, 'message': 'Sale completed successfully'})
    except IntegrityError:
        db.session.rollback()
        return jsonify({'success': False, 'message': 'Integrity error during transaction'}), 400
    except Exception as e:
        db.session.rollback()
        return jsonify({'success': False, 'message': str(e)}), 500


@sales_bp.route('/reports/daily', methods=['GET'])
@limiter.limit("50 per day")
@login_required
def daily_sales_report():
    today = datetime.today().date()  # Get today's date
    sales = Sale.query.filter(db.func.date(Sale.date) == today).all()

    if not sales:
        flash("No sales data available for today", "info")

    return render_template('daily_sales_report.html', sales=sales, today=today)


# Weekly sales report
@sales_bp.route('/reports/weekly', methods=['GET'])
@limiter.limit("50 per week")
@login_required
def weekly_sales_report():
    one_week_ago = datetime.utcnow().date() - timedelta(days=7)
    sales = Sale.query.filter(Sale.date >= one_week_ago).all()
    return render_template('weekly_sales_report.html', sales=sales)


@sales_bp.route('/reports/filter', methods=['POST'])
@login_required
def filter_sales_report():
    start_date_str = request.json.get('start_date')
    end_date_str = request.json.get('end_date')

    # Parse the dates from the request
    start_date = datetime.strptime(start_date_str, '%Y-%m-%d')
    end_date = datetime.strptime(end_date_str, '%Y-%m-%d')

    # Use join to optimize querying sales with items
    sales = Sale.query.filter(Sale.date.between(start_date, end_date)).all()

    # Initialize report data and totals
    report_data = {}
    total_revenue = 0
    total_items_sold = 0
    total_profit = 0

    # Process sales data to gather required information
    for sale in sales:
        sale_items = sale.cart_items  # Get all items in the sale
        total_sale_value = sale.total  # Total revenue for this sale

        for item in sale_items:
            product_name = item.product.name  # Get the product name
            quantity = item.quantity
            unit_price = item.product.price  # Cost price of the product
            cost_price = unit_price * quantity
            profit = total_sale_value - cost_price

            # Group by product name instead of category
            if product_name not in report_data:
                report_data[product_name] = {
                    'total_sold': 0,
                    'total_revenue': 0,
                    'cost_price': 0,
                    'profit': 0,
                }

            # Accumulate the values for each product
            report_data[product_name]['total_sold'] += quantity
            report_data[product_name]['total_revenue'] += total_sale_value
            report_data[product_name]['cost_price'] += cost_price
            report_data[product_name]['profit'] += profit

            # Aggregate totals
            total_revenue += total_sale_value
            total_items_sold += quantity
            total_profit += profit

    # Convert the report data into a list for the JSON response
    report_data_list = [
        {
            'product_name': product,
            'total_sold': data['total_sold'],
            'total_revenue': data['total_revenue'],
            'cost_price': data['cost_price'],
            'profit': data['profit'],
        }
        for product, data in report_data.items()
    ]

    # Return the sales data and aggregated totals in JSON format
    return jsonify({
        'sales': report_data_list,
        'total_revenue': total_revenue,
        'total_items_sold': total_items_sold,
        'total_profit': total_profit
    })
